compiler/util/misc.cpp

```diff
@@ -197,8 +205,52 @@ printDevelErrorHeader(BaseAST* ast) {
   }
 
 
-  if (ast && ast->linenum())
-    fprintf(stderr, "%s:%d: ", cleanFilename(ast), ast->linenum());
+  if( ast && ast->linenum() ) {
+    have_ast_line = 1;
+    filename = cleanFilename(ast);
+    linenum = ast->linenum();
+  } else {
+    have_ast_line = 0;
+    if( !err_print && currentAstLoc.filename && currentAstLoc.lineno > 0 ) {
+      // Use our best guess for the line number for user errors,
+      // but don't do that for err_print (USR_PRINT) notes that don't
+      // come with line numbers.
+      filename = cleanFilename(currentAstLoc.filename);
+      linenum = currentAstLoc.lineno;
+    } else {
+      filename = NULL;
+      linenum = -1;
+    }
+  }
+
+  if (err_user)
+    apologize = 0;
+  else
+    apologize = 1;
+
+  if( apologize ) {
+    fprintf(stderr, " Unfortunately the Chapel compiler has encountered\n"
+                    " an internal error. Please file a bug report that\n"
+                    " includes a program reproducing the problem as well\n"
+                    " as this output. See %s for\n"
+                    " further instructions on filing bug reports.\n"
+                    "\n", help_url);
+    if( !have_ast_line && filename) {
+      // Print out our best guess for the location of an error
+      // if we had no source location
+      fprintf(stderr, " The error may be related to this location:\n"
+                      " %s:%d\n",
+                      filename, linenum);
+    }
```

Comment:
I didn't do that because some cases we are getting the correct line numbers but couldn't provide an AST element easily (e.g. somehow your test with compilerWarning  test/users/vass/type-tests.isSubtype is ending up with ast=NULL and I think it has to do with delayed errors in resolution). Basically, doing it this way minimized the disturbance to other tests, which seemed like a good idea since my goal here is improving core-dump errors vs improving all errors. That can be future work.